server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1200,17 +1216,26 @@ private static Set<String> innerResolve(Context context, List<String> expression
                     continue;
                 }
                 wildcardSeen = true;
-
                 final Map<String, IndexAbstraction> matches = matches(context, expression);
                 if (context.getOptions().allowNoIndices() == false && matches.isEmpty()) {
                     throw indexNotFoundException(expression);
                 }
-                Set<String> expand = expand(context, matches, expression);
-                if (add) {
-                    result.addAll(expand);
-                } else {
-                    result.removeAll(expand);
-                }
+                Collection<String> finalResult = result;
+                expandMatches(context, matches.values(), expression.startsWith("."), expanded -> {
```

Comment:
> Can we call expressions that start with dots implicitly hidden ? 

Not quite. The logic is that wildcard expressions that start with "." will cover hidden indices that also start with ".", even if the request options set `expandWildcardsHidden` to `false`.

> For example Regex.isSimpleMatchPattern(expression) is not checked here as part of the implicitHidden check and hard to tell if the result is like for like.

`Regex.isSimpleMatchPattern(expression)` is assumed `true` at this stage, in the `expandMatches` method (expansion happens only for wildcards).
The intention is to eliminate the `expression` variable from the expansion method. The `expression` is used in matches, but then expanding the matches shouldn't need the `expression` anymore.

Reply:
Tangentially:

Depending on the context, individual wildcards that expand to the empty list will throw exceptions.

But this is not `true` in the case that the wildcard expands to hidden or system indices which are then disregarded (so it's as if the wildcard expanded to nothing, but no error is thrown).
I would consider this a bug, and solving this would actually simplify the code here because the `expansion` method would not be concerned about whether the expression starts with a "." or not.

If you agree this is a bug I would LOVE to fix it.

Reply:
In Security, a wildcard that expands only to hidden/system resources will indeed throw the `indexNotFound` exception, so I would even more lean in treating this behavior as a bug (ie that a wildcard that expands to only hidden/system resources does NOT throw the `indexNotFound` exception).

Reply:
@jakelandis Above I'm proposing that wildcards that ultimately expand to nothing because they only expand to hidden or system indices, that are ultimately not resolved, throw the index not found exception (this is what happens when Security is enabled or when the wildcard expands to no index at all). In a chat, you've asked what happens to the hidden backing indices of data streams.

Because said indices start with `.ds-` a wildcard that accesses them must also start with `.` (or be the `_all` or `*`, which is a special case outside of this discussion), for eg `.ds*`. All wildcards that start with `.`  will match indices that start with `.` even if the indices are hidden. So wildcards for backing indices of datastreams are not concerned about the change.

So, for example, this is what I'm proposing to fix:
 * say there is only a single `index` in the cluster, which is created as a hidden index. `index*/_search?allow_no_indices=false` returns empty result (only when Security is disabled), but `test*/_search?allow_no_indices=false` throws the index_not_found 400 exception.

Reply:
I am pretty sure that is bug.  Per our doc: https://www.elastic.co/guide/en/elasticsearch/reference/current/api-conventions.html#multi-hidden the request should error. 

> allow_no_indices
    (Optional, Boolean) If false, the request returns an error if any wildcard expression, [index alias](https://www.elastic.co/guide/en/elasticsearch/reference/current/aliases.html), or _all value targets only missing or closed indices. This behavior applies even if the request targets other open indices. For example, a request targeting foo*,bar* returns an error if an index starts with foo but no index starts with bar.

In your example, if this was the request
`index*/_search?allow_no_indices=false&expand_wildcards=hidden` then it should return results (and not an error).

Reply:
> In your example, if this was the request
index*/_search?allow_no_indices=false&expand_wildcards=hidden then it should return results (and not an error).

Correct.

I'm going to follow-up with a quick PR to fix this.